lib/utils/README.md

```diff
@@ -2,19 +2,375 @@
+
+2. Even if we disable default constructibility, the ability to perform aggregate initialization of a subset of the `struct`'s fields is still undesirable. For example,
+```cpp
+struct Person {
+  Person() = delete;
+
+  std::string first_name;
+  std::string last_name;
+  int age = 0;
+};
+
+Person p{"donald", "knuth"};
+```
+will compile just fine, but will silently use a nonsensical value of `age`. Even worse, if we were to add an additional field `is_male`
 ```cpp
+struct Person {
+  Person() = delete;
+
+  std::string first_name;
+  std::string last_name;
+  int age = 0;
+  bool is_male = false;
+};
+
+Person p{"donald", "knuth", 85};
+```
+would compile without any errors, but is (as of writing this) incorrect.
+
+3. For product types, `operator==` and `operator!=` are trivial, but still have to be written and maintained, and can easily lead to bugs. For example, 
+```
+struct Person {
+  Person() = delete;
+  Person(std::string const &first_name, 
+         std::string const &last_name, 
+         int age)
+    : first_name(first_name),
+      last_name(last_name),
+      age(age),
+    { }
+
+  friend bool operator==(Person const &lhs, Person const &rhs) {
+    return lhs.first_name == rhs.first_name 
+      && lhs.last_name == rhs.last_name
+      && lhs.age == rhs.age;
+  }
+
+  friend bool operator!=(Person const &lhs, Person const &rhs) {
+    return lhs.first_name != rhs.first_name 
+      || lhs.last_name != rhs.last_name
+      || lhs.age != rhs.age;
+  }
 
+  std::string first_name;
+  std::string last_name;
+  int age;
+};
 ```
+If we take the previous example of adding an additional `is_male` field to `Person`, it can be easy to miss a location, leading to incorrectness. 
+For example, we could quite easily end up with
+```cpp
+struct Person {
+  Person() = delete;
+  Person(std::string const &first_name, 
+         std::string const &last_name, 
+         int age,
+         bool is_male)
+    : first_name(first_name),
+      last_name(last_name),
+      age(age),
+      is_male(is_male)
+    { }
+
+  friend bool operator==(Person const &lhs, Person const &rhs) {
+    return lhs.first_name == rhs.first_name 
+      && lhs.last_name == rhs.last_name
+      && lhs.age == rhs.age
+      && lhs.is_male == rhs.is_male;
+  }
+
+  friend bool operator!=(Person const &lhs, Person const &rhs) {
+    return lhs.first_name != rhs.first_name 
+      || lhs.last_name != rhs.last_name
+      || lhs.age != rhs.age;
+      // oops, forgot to update here. Have fun debugging :P
+  }
+
+  std::string first_name;
+  std::string last_name;
+  int age;
+  bool is_male;
+};
+```
+and for product types with more fields this grows increasingly tedious to write and maintain. 
+
+4. Hashing: hashing for product types is also relatively trivial, as long as each of the fields is hashable. But again, we have to do a bunch of extra work to specify this, and this work has to be done for each product type in the codebase.
```

Comment:
Reword to "Hashing product types is relatively trivial, long as each of the fields..."